src/number_theory/legendre_symbol/gauss_sum.lean

```diff
@@ -0,0 +1,315 @@
+               of_add_zero, monoid_hom.map_one, mul_one],
+    exact hχ.sum_eq_zero.symm, },
+  { -- case `b ≠ 0`
+    refine (fintype.sum_bijective _ (mul_left_bijective₀ b hb) _ _ $ λ x, _).symm,
+    rw [mul_assoc, mul_comm x, ← mul_assoc, mul_inv_cancel hb, one_mul, mul_sub, mul_one] },
+end
+
+/-- We have `gauss_sum χ ψ * gauss_sum χ⁻¹ ψ⁻¹ = fintype.card R`
+when `χ` is nontrivial and `ψ` is primitive (and `R` is a field). -/
+lemma gauss_sum_mul_gauss_sum_eq_card  {χ : mul_char R R'} (hχ : is_nontrivial χ)
+  {ψ : add_char R R'} (hψ : is_primitive ψ) :
+  gauss_sum χ ψ * gauss_sum χ⁻¹ ψ⁻¹ = fintype.card R :=
+begin
+  simp only [gauss_sum, inv_mul_shift, mul_shift, monoid_hom.coe_comp,
+             function.comp_app, add_monoid_hom.to_multiplicative_apply_apply, finset.mul_sum,
+             add_monoid_hom.coe_mul_left, neg_mul, one_mul, of_add_neg, of_add_to_add],
+  simp_rw [finset.sum_mul, mul_mul_mul_comm, inv_apply' χ, ← map_mul χ, ← map_mul ψ],
+  convert_to ∑ y x, χ (x * y⁻¹) * ψ (of_add (x - y)) = _,
+  { simp_rw sub_eq_add_neg, refl },
+  simp_rw gauss_sum_mul_aux hχ ψ,
+  rw [finset.sum_comm],
+  classical, -- to get `[decidable_eq R]` for `gauss_sum_mul_aux₂`
+  simp_rw [← finset.mul_sum, sum_mul_shift _ _ hψ, sub_eq_zero, mul_ite, mul_zero],
+  rw [finset.sum_ite_eq' finset.univ (1 : R)],
+  simp only [finset.mem_univ, map_one, one_mul, if_true],
+end
+
+/-- When `χ` is a nontrivial quadratic character, then the square of `gauss_sum χ ψ`
+is `χ(-1)` times the cardinality of `R`. -/
+lemma gauss_sum_sq  {χ : mul_char R R'} (hχ₁ : is_nontrivial χ) (hχ₂ : is_quadratic χ)
+  {ψ : add_char R R'} (hψ : is_primitive ψ) :
+  (gauss_sum χ ψ) ^ 2 = χ (-1) * fintype.card R :=
+begin
+  rw [pow_two, ← gauss_sum_mul_gauss_sum_eq_card hχ₁ hψ, hχ₂.inv, mul_rotate'],
+  congr,
+  rw [mul_comm, ← gauss_sum_mul_shift _ _ (-1 : Rˣ), inv_mul_shift],
+  refl,
+end
+
+end gauss_sum_prod
+
+/-!
+### Gauss sums and Frobenius
+-/
+
+section gauss_sum_frob
+
+variables {R : Type u} [comm_ring R] [fintype R] {R' : Type v} [comm_ring R']
+-- We assume that the target ring `R'` has prime characteristic `p`.
+variables (p : ℕ) [fp : fact p.prime] [hch : char_p R' p]
+include fp hch
+
+/-- When `R'` has prime characteristic `p`, then the `p`th power of the Gauss sum
+of `χ` and `ψ` is the Gauss sum of `χ^p` and `mul_shift ψ p`. -/
+lemma gauss_sum_frob (χ : mul_char R R') (ψ : add_char R R') :
+  gauss_sum χ ψ ^ p = gauss_sum (χ ^ p) (ψ ^ p) :=
+begin
+  rw [← frobenius_def, gauss_sum, gauss_sum, map_sum],
+  simp_rw [pow_apply' χ fp.1.pos, map_mul, frobenius_def],
+  refl,
+end
+
+/-- For a quadratic character `χ` and when the characteristic `p` of the target ring
+is a unit in the source ring, the `p`th power of the Gauss sum of`χ` and `ψ` is
+`χ p` times the original Gauss sum. -/
+lemma quad_gauss_sum_frob (hp : is_unit (p : R)) {χ : mul_char R R'} (hχ : is_quadratic χ)
```

Comment:
Would it make sense to put this in the `is_quadratic` namespace?

Reply:
Yes, I think so. Will do.

Reply:
...or rather, `mul_char.is_quadratic`, since this allows for dot notation like `hχ.quad_gauss_sum_frob _ hp`.